new service factory: CPS (Cloud Pub/Sub)

dhrgit opened the pull request:
This commit adds a new service factory, whereby messages are delivered by publishing to a Cloud Pub/Sub topic.

torsm commented:
Thanks for the PR Dan, could you re-run generate_go_py_protos.sh? a diff in generated proto files is blocking the merge atm

gnoack commented on fleetspeak/src/server/cpsservice/service_test.go:
```diff
@@ -0,0 +1,137 @@
+}
+
+func TestProcessMessage(t *testing.T) {
+
```
inconsistency: this is the only function that has an extra newline at the start

dhrgit replied:
Fixed.

gnoack commented on fleetspeak/src/server/cpsservice/service_test.go:
```diff
@@ -0,0 +1,137 @@
+	sub *pubsub.Subscription
+}
+
+func createTestEnv(t *testing.T) testEnv {
```
If a test helper calls t.Fatalf() or passes the t object onwards, it should call t.Helper() at the top. That makes sure that the failure message is attributed to the surrounding test, rather than to the line within the helper function.

See https://google.github.io/styleguide/go/decisions#must-functions (second example) or https://blog.gnoack.org/post/go-tests/ for some examples.

dhrgit replied:
Done.

torsm commented on fleetspeak/src/server/cpsservice/service.go:
```diff
@@ -0,0 +1,78 @@
+
+// CpsService is a service.Service which publishes all received
+// messages to a Cloud Pub/Sub topic.
+type CpsService struct {
```
Acronyms should be capitalized (-> `CPSService`)

dhrgit replied:
Done.

torsm commented on fleetspeak/src/server/cpsservice/service.go:
```diff
@@ -0,0 +1,78 @@
+// the configuration in cfg.
+func Factory(cfg *spb.ServiceConfig) (service.Service, error) {
+	conf := cpspb.Config{}
+	if err := cfg.Config.UnmarshalTo(&conf); err != nil {
```
Proto fields should be accessed using getters
```suggestion
	if err := cfg.GetConfig().UnmarshalTo(&conf); err != nil {
```

dhrgit replied:
Done.

torsm commented on fleetspeak/src/server/cpsservice/service.go:
```diff
@@ -0,0 +1,78 @@
+		return nil, err
+	}
+
+	client, err := pubsub.NewClient(context.Background(), conf.Project)
```
```suggestion
	client, err := pubsub.NewClient(context.Background(), conf.GetProject())
```

dhrgit replied:
Done.

torsm commented on fleetspeak/src/server/cpsservice/service.go:
```diff
@@ -0,0 +1,78 @@
+		return nil, err
+	}
+
+	s := CpsService{topic: client.Topic(conf.Topic)}
```
```suggestion
	s := CpsService{topic: client.Topic(conf.GetTopic())}
```

dhrgit replied:
Done.

dhrgit commented:
> Thanks for the PR Dan, could you re-run generate_go_py_protos.sh? a diff in generated proto files is blocking the merge atm

Sure - I added this as a new commit since it touched a bunch of proto files. Let me know if you'd rather I squashed it.

torsm commented:
Oh no it seems that it's generating the go proto files with a different version than expected 🤔 
Granted, this part of fleetspeak development is frustratingly tricky...

the github actions are setup with the following versions:
```
          go install google.golang.org/protobuf/cmd/protoc-gen-go@v1.28.0
          go install google.golang.org/grpc/cmd/protoc-gen-go-grpc@v1.2
```

were you running the script from a python virtualenv specifically for fleetspeak, where fleetspeak_python is installed? That should come with the correct versions for everything

dhrgit commented:
> were you running the script from a python virtualenv specifically for fleetspeak, where fleetspeak_python is installed? That should come with the correct versions for everything

Ah, I see. Running from a `fleetspeak_python` venv seems to have done the trick: only the cpsservice proto was updated.
